### PR TITLE
[SelectionDAG] Add initial support for nneg flag on ISD::ZERO_EXTEND.

### PR DESCRIPTION
**Pull request by topperc**
This adds the nneg flag to SDNodeFlags and the node printing code. SelectionDAGBuilder will add this flag to the node if the target doesn't prefer sign extend.

A future RISC-V patch can remove the sign extend preference from SelectionDAGBuilder.

I've also added the flag to the DAG combine that converts ISD::SIGN_EXTEND to ISD::ZERO_EXTEND.

### CODE REVIEW

**Review by nikic — COMMENTED** [review]
You need to update the SDAG canCreateUndefOrPoison() implementation to account for the new flag.

This generally looks ok to me, but I'll leave it to people who do more backend work to decide whether having this flag in SDAG is worthwhile. Given how target-specific SDAG is, I could see the argument for just using SEXT based on target preference and not adding the complexity of additional poison flags.

**Comment by preames** [comment]
Do we need to add the flag to the zext_in_reg nodes?   This question is complicated by the fact that comments in DAGCombine mention such a node, but it doesn't appear to be in the ISDOpcodes.h list.  

Can you update the comment for ZERO_EXTEND in ISDOpcodes.h to mention the flag and the implied poison semantics?

For context, we're definitely going to need this for RISCV.  At the moment, I set us up to prefer SIGN_EXTEND based on the target preference, and I see some cases in benchmarks where this hurt codegen because we missed a zext fold or zeroing bits is easier in a particular case (say, and with smallish constant).

**Comment by topperc** [comment]
> Do we need to add the flag to the zext_in_reg nodes? This question is complicated by the fact that comments in DAGCombine mention such a node, but it doesn't appear to be in the ISDOpcodes.h list.

zext_in_reg is ISD::AND with a trailing ones mask.

**Review by preames — APPROVED** [review]
LGTM

File: llvm/include/llvm/CodeGen/ISDOpcodes.h [hunk 1]
```diff
@@ -773,7 +773,10 @@ enum NodeType {
   /// into new bits.
   SIGN_EXTEND,
 
-  /// ZERO_EXTEND - Used for integer types, zeroing the new bits.
+  /// ZERO_EXTEND - Used for integer types, zeroing the new bits. Can carry
+  /// the NonNeg SDNodeFlag to indicate that the input is known to be
+  /// non-negative. If the flag is present and the input is negative, the reuslt
```
Comment (nikic) [inline llvm/include/llvm/CodeGen/ISDOpcodes.h:778]:
```suggestion
  /// non-negative. If the flag is present and the input is negative, the result
```

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp [hunk 2]
```diff
@@ -5037,6 +5036,10 @@ bool SelectionDAG::canCreateUndefOrPoison(SDValue Op, const APInt &DemandedElts,
   case ISD::BUILD_PAIR:
     return false;
 
+  // Matches hasPoisonGeneratingFlags().
+  case ISD::ZERO_EXTEND:
+    return ConsiderFlags && Op->getFlags().hasNonneg();
```
Comment (nikic) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:5041]:
```suggestion
    return ConsiderFlags && Op->getFlags().hasNonNeg();
```

**Review by nikic — COMMENTED** [review]
Should also update SimplifyDemanded to drop that flag.

**Comment by nikic** [comment]
> > Do we need to add the flag to the zext_in_reg nodes? This question is complicated by the fact that comments in DAGCombine mention such a node, but it doesn't appear to be in the ISDOpcodes.h list.
> 
> zext_in_reg is ISD::AND with a trailing ones mask.

There is ZERO_EXTEND_VECTOR_INREG though, so I guess the question applies to that one.

**Pull request merged**